action.yml

```diff
@@ -0,0 +1,50 @@
+name: Sync Salmon Directory
+description: Syncs a directory to a Google Cloud bucket using rclone.
+author: 'deephaven'
+inputs:
+  source:
+    required: true
+    type: string
+    description: 'The source directory to sync.'
+  destination:
+    required: true
+    type: string
+    description: 'The destination directory to sync. Relative to the bucket.'
+  project_number:
+    required: true
+    type: string
+    description: 'The Google Cloud project number.'
+  bucket:
+    required: true
+    type: string
+    description: 'The Google Cloud bucket to sync to.'
+  credentials:
```

Comment:
I wonder if these can be abstracted away to just read an env variable that we expect in the action runner. Would simplify calling the action a bit. Not a big deal if they can't, but I figure basically everywhere we'll be using the same values here.

Unless there's a use case we want to serve out of multiple buckets

Reply:
I definitely had that thought. I think it's also a nice idea because then the credentialing is independent of usage in case we change providers again. 
The reason I didn't do it initially was security considerations. This is a public action and we'd need to make sure everything was configured properly. So I'm totally in support of doing it, I would just want some reassurance that this is something we can do safely from someone that knows more than I do...

Reply:
I think you would be safe if it just referenced `secrets.TOKEN` or something like that. The token would be available to the Deephaven org so any repo running should have it in its secrets already. Anybody outside of Deephaven using this action (for whatever reason) shouldn't have access to the secret and the action would fail if the secret wasn't defined

I'm not sure if a reusable action needs anything extra. You might still have to provide it as an input to the action. I see examples from GitHub that implies secrets are safe to pass to a reusable action and not get exposed.

Another option would be to make the action private and share it w/ the org. Would still put the token in a secret. https://docs.github.com/en/actions/sharing-automations/sharing-actions-and-workflows-with-your-organization

Reply:
I would at least try to remove the bucket and project number from the inputs. The token might be needed, but we can at least make it an org level secret.

Would anybody ever change the bucket or project number? I'm not sure if those need to be private or not. Can probably ask in devops. I'd think that data (at least bucket name) would get exposed to end users anyway on the docs site so they can access images.

Reply:
The private part at least I can't do I think because (from that link)
"Any actions or reusable workflows stored in the private repository can be used in workflows defined in other private repositories owned by the same organization. Actions and reusable workflows stored in private repositories cannot be used in public repositories."
And this would be used in both public and private repos

I don't think bucket and project number are strictly secrets, but the less broadcast about our config the better, even if you could seek them out.

My understanding of secrets in this case is that this composite action doesn't have access to secrets by default from the calling workflow and they need to be passed in as inputs (like I currently have) or we somehow filter so only deephaven projects can call this workflow but then this composite workflow has access to the secrets somehow (from the org)

Perhaps, to connect all this, the best way forward is to only take the credential, verify the credential matches the expected credential, then continue. Then whether or not the bucket and project number are secrets isn't important.  If we ever change the cloud provider, it's also much easier, because we can just use that new credential. I think I'll go that route.

Reply:
To link in that convo here, James says project number is best to keep secret, bucket is not a big deal, but I don't see much point in exposing the bucket if we don't have to either.
Also, after playing with this for a bit I think it's actually not at all possible to use this repo's secrets? A bit disappointing but I'll revert all that. I think we'll just have to pass these things in unfortunately.

Reply:
Ya I think actions have to be passed secrets. Which makes sense for a marketplace action where it could choose to dump secrets and send it wherever which would be awful.

The bucket name could just be coded into the action if you wanted to cut down on inputs, but if the readme just had a copy/paste with inputs then that's fine. There's like a 99% chance the bucket will be part of asset URLs anyway I think? Or maybe we're not serving images directly from the bucket, but we might want to.

Reply:
added a copy paste example, removing the bucket wouldn't be any less tedious with the other requirements so just going with that